### ISSUE TITLE
Retraining detector

### ISSUE THREAD

**Issue by abhigarg** [tancik/StegaStamp#5]
I did some testing and found that it requires a white border around the image containing the hidden code for the detector to work. For my use-case, I need the detector to work even without the white border. So I will have to retrain the detector.

For that I am following the other github [repo](https://github.com/GeorgeSeif/Semantic-Segmentation-Suite) but it seems to be causing some issues with the input size of model. 

Has anyone tried this?

**Comment by CYang0515** [comment]
Please, I have confusion that the paper is how to build the dataset to train the detector. The paper said the method is  " randomly transformed StegaStamps embedded into high resolution images sampled from DIV2K ". However the StegaStamps is sampled from the high resolution or isn't related to the high resolution.

**Comment by tancik** [comment]
@abhigarg The decoder we trained used white borders, however the detector did not. You should be able to use the pretrained detector with a newly trained decoder that uses other colored backgrounds (during the augmentation step). In regards to your issue with the input size, we did not run into this problem. We would recommend creating an issue in that repo.

@CYang0515 The DIV2K dataset provides high resolution images. The stegastamps we generate are all 400x400 px. These are randomly scaled and placed on the high resolution images.

**Comment by abhigarg** [comment]
@tancik Where can I find  the newly trained decoder?

**Comment by tancik** [comment]
I was inferring that it is possible to retrain the decoder with different backgrounds. The readme outlines the steps needed to retrain the model.

**Comment by abhigarg** [comment]
@tancik Thanks for information. I am wondering is it possible to make the model input size independent. Current model takes only 400 x 400 image as input. If I have an original image say 720 x 1280, then I will have to reduce it to 400 x 400 and then hide a message but that will reduce the pic quality. Another way is to crop out a 400 x 400 from the original and hide qr in it then put it back in original image. But then I can see the borders i.e. the 400 x 400 does not merge smoothly with the rest of the pic.

**Comment by tancik** [comment]
That is one of the current limits to the system. Embedding codes across the entire image is not something we explore in this work. We think this is an exciting extension and we welcome input / solutions from the community.
Here are a couple of ideas that I have though about:
+ Process image as 400x400 crop but only save the residual. Rescale the residual to the target dimension and add it to the original image.
+ To embed into an existing image with reduced borders you can try increasing the l2 loss at the edges. Note that the lack of edges will likely make detection more difficult, so the detection algorithm may need to be updated.

**Comment by abhigarg** [comment]
@tancik Thanks for the idea of rescaling the residual to target dimension. I am trying in my python code but the results are far from expectation. I am not sure if I am coding it correctly. I am wondering if you can take a look at how I am rescaling and adding to the original. I also try to add residual for 400 x 400 and add it to 400 x 400 scaled input image.

#encoded image and residual for 400 x 400 scaled input image
`hidden_img, residual = sess.run([output_stegastamp, output_residual],feed_dict=feed_dict)`
`rescaled = (hidden_img[0] * 255).astype(np.uint8)`
`rescaled_residual = (residual[0] * 255).astype(np.uint8)`

#add residual to 400 x 400 scaled input image
`test_img = cv2.add(deepcopy(scaled_target_image_400x400), rescaled_residual)`
`test_img_bgr = cv2.cvtColor(test_img, cv2.COLOR_RGB2BGR)`

#scaling of residual for 400 x 400 input
`h_orig_img, w_orig_img = orig_img.shape[:2]`
`residual_scaled = cv2.resize(residual[0],(w_orig_img,h_orig_img))`
`residual_scaled = (residual_scaled * 255).astype(np.uint8)`

#add scaled residual to original target input image
`orig_output_rgb = cv2.add(deepcopy(orig_img_rgb), residual_scaled)`
`orig_output = cv2.cvtColor(orig_output_rgb, cv2.COLOR_RGB2BGR)`

**Comment by CYang0515** [comment]
@tancik  The  stegastamps come from a different dataset(mirflickr) with the high resolution images(DIV2K).  When training a detector, the randomly scaled stegastamps are padded into high resolution images. The synthetic dataset is obvious, why is it so difficult to train the detector that needs a larger l2 loss at image edges?

**Comment by abhigarg** [comment]
@tancik 
the resizing residual to original image seems to be working. I used `np.add()` instead of `cv2.add()`.
Also, after adding to original, I had to truncate all values > 255 to 255 and all values < 0 to 0. now the results looks much alike. Thanks for the wonderful idea !

**Comment by tancik** [comment]
@CYang0515 I was refering to the encoder having a larger l2 loss at the edges so that the residual would have a smaller magnitude at the edges. Then if you pasted the residual into a patch in the larger image, it would be harder to detect as there are not sharp edges.

@abhigarg Great to hear that it works!

**Comment by CYang0515** [comment]
@tancik 
![D1](https://user-images.githubusercontent.com/45357137/60764360-e5b95380-a0ba-11e9-86e8-6dc0b61a7bee.png)
 If the patch encoding the residual, is a part of the larger image, it wouldn't  have sharp edges when pasting the residual into the patch. But if the content of patch isn't related the larger image,  it always has a sharp edges. The synthetic dataset training the detector is like this, the center image is stegastamps coming from mirflickr and larger image coming from DIV2K. there are any problems? when training the detector, the edges by l2 loss have any impact?

**Comment by tancik** [comment]
@CYang0515 Ahh I understand the confusion now. The detector can be trained in two different ways:
1. Place random mirflickr stegastamps into the high resolution image. In this case the edge loss does not impact training the training of the detection model. This is the case you describe above.
2. Extract crops from high resolution image, apply stegastamp encoding to crop, reinsert encoded crop back into the image from where they were extracted. In this case the residual at the edges will make the stagastamps more visible. Increasing the edge loss reduces this effect, but makes it harder for the detector to locate.

**Comment by CYang0515** [comment]
@tancik Oh, I get it. And the model that you provide, is which way to train? Have you try both ways?

**Comment by tancik** [comment]
@CYang0515 The provided trained detector uses method 1 from above. I briefly experimented with method 2 above and got mediocre results. however, I still think there is a fair bit more experimentation in this area that could be done. Since the detector was not the focus of this project, we did not invest too much time into it.

**Comment by goldfish22** [comment]
@tancik 
Hello,I am a little confusion about the detector.
I can use the encoder and the decoder to hide and reveal the information.
But when I use the detector.py to detect and decode the information,I input a picture with hidden information (ie.the output image of encoder.py) to the script, but it can't detect and decode the hidden information, why was that?I also try to photograph the picture with hidden information, it also fail to detect the stegastamp and has no decode results.

**Comment by 18360939479** [comment]
Hello, I can decode the screenshot of the picture, but after taking a picture on my mobile phone, I decode the picture and it fails. that‘s why?

**Comment by tancik** [comment]
@goldfish22, @18360939479 It sounds like the detector may be failing to detect the correct crop. I would try with the `--visualize_detector` argument to visualize the crops found by the segmentation algorithm. If no crops are found, or if the crops are very off, the decoder will not work. The detector is an off-the-shelf segmentation network, depending on your usecase, you may need a more specialized detector.

**Comment by kzpsycheuu** [comment]
@goldfish22 Have you find a way to detect the information in a single image?
